### ISSUE TITLE
Calculating number of partitions / counts should not require full deserialization of root file

### ISSUE THREAD

**Issue by lgray** [spark-root/laurelin#52]
Also something we discussed in meeting of two weeks ago.

Returning the number of partitions in an rdd, and similarly querying the number of rows, for a root file should not require the full deserialization of the root files and can likely be made significantly faster.

This is primarily a UX improvement.

**Comment by PerilousApricot** [comment]
Not sure how I missed this mail, but thanks!

**Comment by PerilousApricot** [comment]
I should be clear, though -- the "full" file isn't deserialized. To open a TTree, you load TFile & friends (small header, the TDictionary and the Streamers)

Unfortunately, the problem is that the streamers end up with something that looks like this (in python for berevity)

```python
ttree_streamer = [
    TStreamerBase("foo"), # superclass of ttree
    TStreamerString("fName"),
    TStreamerObject("foo"),
    TStreamerObject("bar"), 
]
```

So, when you're trying to deserialize the TTree, you would normally do something like
```python
fh = open("file", "r").seek(start_of_ttree)
tree = dict()
for streamer_obj in ttree_streamer:
    # this function updates dict with data from the file, deserialized by the particular streamer and
    # moves the filehandle forward to the position of the next streamable element
    streamer_obj.deserialize(tree, fh)
```

The problem is that each streamer element only knows it's own length *after* deserialization -- each bit could possibly be variable-length, afterall. But that means that there's no (good) way to immediately seek to `bar` without at least reading some of `foo`'s info. At the very least, we have to read and deserialize some internal metadata about `foo` to know how much to skip.

Now, if `foo` was a particularly complex/nested object, it would be a win to jump through the hoops to do a "seek instead of deserialize", but unfortunately, the complex/nested objects are the TBranch* and TLeaf* objects :( DO NOT CLICK IN YOUR BROWSER IT IS 30 MEGS AND WILL CHOKE -- look at the following dump of the TTree data structure of a miniaod file:

http://mirror.accre.vanderbilt.edu/spark/dump_3g.txt

Most of the nested structure is TBranch/TLeaf, and we can't prune that because we need them. We can do something like skipping over fixed length-fields for uninteresting data (like `fBits`), but I don't think it'll be a giant win

**Comment by lgray** [comment]
@PerilousApricot so, trying to scale this up to a full analysis, a major time sink is that we need to deserialize the first file in a file list at the driver in order to establish something... a schema I guess?

Anyway, this is impressively slow when you start having 52 datasets to analyze, so we really need another way to do this.

**Comment by PerilousApricot** [comment]
Can you document here what code you're executing, what resources you're running on, etc.. to have that info around?

**Comment by lgray** [comment]
Found the problem. Eager deserialization of baskets. Have it patched out and made lazy in local code. Testing.

**Comment by PerilousApricot** [comment]
baskets or basket metadata?

**Comment by lgray** [comment]
I assume getBaskets() gets baskets.

**Comment by lgray** [comment]
Anyway this is going at scale of a full analysis 52 datasets of varying size, with the FNAL spark cluster using 200 cores.

**Comment by PerilousApricot** [comment]
That's the basket metadata, not the actual basket content

**Comment by lgray** [comment]
Executing this analysis in python:
https://github.com/nsmith-/coffeandbacon/blob/master/analysis/boostedHbbProcessor.py

It loads ~100 columns or so. 

As things are in 0.3.0 (note that master presently does not function, incorrect array sizes are yielded) the code here will get executed at the driver:
https://github.com/spark-root/laurelin/blob/laurelin-0.3.0/src/main/java/edu/vanderbilt/accre/laurelin/root_proxy/TBranch.java#L94-L120

and for large files with many baskets is very slow. Furthermore, nothing is done with it on the driver. 
It's better to delay this to when it's needed.

**Comment by PerilousApricot** [comment]
`note that master presently does not function, incorrect array sizes are yielded` - can you make an issue for that?

**Comment by lgray** [comment]
Yep, I will. Going to try to get a plot out with a realistic dataset first, no matter how long it takes.

The partition per basket thing also seems to be a nasty show stopper. Overparallelizes on the cluster, thrashes xrootd.

**Comment by lgray** [comment]
Ah, the next thing appears to be that select() and cache() are not lazy operations with this data source. There seems to be some issue in the data source with operation fusing, don't see any optimization.

Digging further.

**Comment by lgray** [comment]
The other hypothesis is that the collect() used in the partition helper is somehow gumming up the works here since it is an action. Perhaps there's a better way to formulate that?

**Comment by PerilousApricot** [comment]
It's a bit hard to follow here, let's move to slack.
